### PR TITLE
Cant close case with no products

### PR DESCRIPTION
**Pull request by KyleMacPherson**
https://regulatorydelivery.atlassian.net/jira/software/projects/PST1/boards/53?selectedIssue=PST1-329

## Description
This changes do not allow users to close a case without any products, instead the user can delete the case.

## Checklist:
- [ ] Have you documented your changes in the pull request description?
- [ ] Does the change present any security considerations?
- [ ] Is any gem functionality overloaded? Eg: Devise controller methods being overloaded.
- [ ] Has acceptance criteria been tested by a peer?
- [ ] Has the CHANGELOG been updated? (If change is worth telling users about.)

### General testing (author)
- [ ] Test without JavaScript
- [ ] Test on small screen

### Accessibility testing (author)
- [ ] Works keyboard only
- [ ] Tested with one screen reader
- [ ] Zoom page to 400% - content still visible
- [ ] Disable CSS - does content make sense and appear in a logical order?

### CODE REVIEW

File: app/controllers/investigations/status_controller.rb [hunk 1]
```diff
@@ -14,6 +14,8 @@ def change_case_status(new_status:, template:, flash:)
       investigation = Investigation.find_by!(pretty_id: params[:investigation_pretty_id])
       authorize investigation, :change_owner_or_status?
 
+      return redirect_to cannot_close_investigation_path(investigation) if investigation.products.count.zero? && new_status == "closed"
```
Comment (stdstaff) [inline app/controllers/investigations/status_controller.rb:17]: Is there a benefit to `investigation.products.count.zero?` over `investigation.products.none?`

**Review by stdstaff — APPROVED** [review]

File: app/forms/delete_investigation_form.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class DeleteInvestigationForm
+  include ActiveModel::Model
+  include ActiveModel::Attributes
+  include ActiveModel::Serialization
+
+  attribute :deleting_user
+  attribute :investigation
+
+  validates :deleting_user, presence: true
```
[the remark this answers is not in the document]
Comment (KyleMacPherson) [reply app/forms/delete_investigation_form.rb]: Done 👍

File: app/models/investigation.rb [hunk 3]
```diff
@@ -80,6 +81,7 @@ class Investigation < ApplicationRecord
   has_many :incidents
   has_many :unexpected_events
 
+  default_scope { where(deleted_at: nil) }
```
[the remark this answers is not in the document]
Comment (KyleMacPherson) [reply app/models/investigation.rb]: Done 👍

File: app/models/investigation.rb [hunk 4]
```diff
@@ -80,6 +81,7 @@ class Investigation < ApplicationRecord
...
   scope :not_private, -> { where(is_private: false) }
 
   redacted_export_with :id, :complainant_reference, :coronavirus_related, :created_at, :custom_risk_level,
```
[the remark this answers is not in the document]
Comment (KyleMacPherson) [reply app/models/investigation.rb]: Done 👍

File: db/migrate/20221130153756_add_deleted_at_and_deleted_by_to_investigations.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class AddDeletedAtAndDeletedByToInvestigations < ActiveRecord::Migration[6.1]
+  def change
+    # rubocop:disable Rails/BulkChangeTable
+    add_column :investigations, :deleted_at, :datetime
```
[the remark this answers is not in the document]
Comment (KyleMacPherson) [reply db/migrate/20221130153756_add_deleted_at_and_deleted_by_to_investigations.rb:5]: Done 👍

File: app/services/delete_investigation.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+class DeleteInvestigation
+  include Interactor
+
+  delegate :investigation, :deleted_by, to: :context
+
+  def call
+    context.fail!(error: "No investigation supplied") unless investigation.is_a?(Investigation)
+    context.fail!(error: "Cannot delete investigation with products") unless investigation.products.count.zero?
```
[the remark this answers is not in the document]
Comment (KyleMacPherson) [reply app/services/delete_investigation.rb:9]: Done 👍

**Comment by KyleMacPherson** [comment]
> We should also add specs which cover the business logic pertaining to deleted cases, e.g:
> 
> * Are they surfaced in search results
> * Are they shown in the totals in the logs used for the Kibana dashboard (`LogDbMetricsJob`)
> * Are they shown in the Cases tab on the product page

Added tests to ensure they're not surfaced in search results, or filtered results. We can't delete cases with products so they won't be shown on the cases tab of the product page and we don't have any tests for the LogDbMetricsJob and I am not sure if we want to show deleted cases, or create a new deleted cases metric? I think that can be tackled separately

**Pull request merged**